bt_ddos_shield/state_manager.py

```diff
@@ -1,78 +1,473 @@
+            self.current_miner_shield_state = self._load_state_from_storage()
 
-    current_miner_shield_state: MinerShieldState
+        return self.current_miner_shield_state
+
+    @abstractmethod
+    def add_validator(self, validator_hotkey: Hotkey, validator_public_key: PublicKey, redirect_address: Address):
+        """
+        Add validator together with his public key and address (created for him) redirecting to Miner server.
+        """
+        pass
 
     @abstractmethod
-    def load_state(self):
+    def update_validator_public_key(self, validator_hotkey: Hotkey, validator_public_key: PublicKey):
         pass
 
     @abstractmethod
-    def save_state(self):
+    def add_banned_validator(self, validator_hotkey: Hotkey):
         pass
 
     @abstractmethod
-    def ban_validator(self, validator_hotkey: Hotkey):
+    def remove_validator(self, validator_hotkey: Hotkey):
         """
-        Add validator to the list of banned validators.
-
-        Args:
-            validator_hotkey: The hotkey of the validator.
+        Remove validator from the sets of known validators and active addresses.
         """
         pass
 
     @abstractmethod
-    def remove_validator(self, validator_hotkey: Hotkey):
-        """
-        Remove validator from the lists of banned validators or active addresses.
+    def set_manifest_address(self, manifest_address: Address):
+        pass
 
-        Args:
-            validator_hotkey: The hotkey of the validator.
+    @abstractmethod
+    def update_address_manager_state(self, key: str, value: Optional[str]):
         """
+        Update address manager state (key -> value). If value is None, remove key from state.
+        """
+        pass
+
+    @abstractmethod
+    def add_address_manager_created_object(self, obj_type: str, obj_id: str):
         pass
 
     @abstractmethod
-    def add_address(self, validator_hotkey: Hotkey, address: Address):
+    def del_address_manager_created_object(self, obj_type: str, obj_id: str):
+        pass
+
+    @abstractmethod
+    def _load_state_from_storage(self) -> MinerShieldState:
+        pass
+
+    def _update_state(self,
+                      known_validators: Optional[dict[Hotkey, PublicKey]] = None,
+                      banned_validators: Optional[dict[Hotkey, datetime]] = None,
+                      validators_addresses: Optional[dict[Hotkey, Address]] = None,
+                      manifest_address: Optional[Address] = None,
+                      address_manager_state: Optional[dict[str, str]] = None,
+                      address_manager_created_objects: Optional[dict[str, frozenset[str]]] = None):
+        """
+        Create new updated state with given new values and set it as current state. If value for field is None,
+        it is copied from current state.
+        """
+        self.current_miner_shield_state = \
+            MinerShieldState(dict(self.current_miner_shield_state.known_validators)
```

Comment:
use or
```
self.current_miner_shield_state = MinerShieldState(
    known_validators or dict(self.current_miner_shield_state.known_validators),
    banned_validators or dict(self.current_miner_shield_state.banned_validators),
    validators_addresses or dict(self.current_miner_shield_state.validators_addresses),
    manifest_address if manifest_address is not None else self.current_miner_shield_state.manifest_address,
    address_manager_state or dict(self.current_miner_shield_state.address_manager_state),
    address_manager_created_objects or dict(self.current_miner_shield_state.address_manager_created_objects),
)
```

Reply:
These dict objects can be empty or None and we need to treat this differently. I changed to 'or' only for manifest_address.